src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets

```diff
@@ -13,24 +13,50 @@
   <!--
     This target hooks into the dotnet CLI publish pipeline. That pipeline has
     a target called ComputeFilesToPublish which produces the ItemGroup
-    ResolvedFileToPublish based on the inputs of @(IntermediateAssembly)
-    and @(ResolvedAssembliesToPublish). We modify those two item groups
-    to control what gets published after NativeAOT optimizes the application.
+    ResolvedFileToPublish. We modify this item group to control what gets
+    published after NativeAOT optimizes the application.
   -->
   <Target Name="ComputeLinkedFilesToPublish"
-          BeforeTargets="ComputeResolvedFilesToPublishList"
+          AfterTargets="ComputeResolvedFilesToPublishList"
           DependsOnTargets="_ComputeAssembliesToCompileToNative;LinkNative">
 
     <ItemGroup>
-      <_ResolvedCopyLocalPublishAssets Remove="@(_AssembliesToSkipPublish)" />
-      <_ResolvedCopyLocalPublishAssets Include="@(_LinkedResolvedAssemblies)" />
-      <_DebugSymbolsIntermediatePath Remove="@(_DebugSymbolsIntermediatePath)" />
+      <ResolvedFileToPublish Remove="@(_AssembliesToSkipPublish)" />
+      <!-- dotnet CLI produces managed debug symbols, which we will replace with native symbols instead -->
+      <ResolvedFileToPublish Remove="@(_DebugSymbolsIntermediatePath)" />
+      <!-- replace apphost with binary we generated during native compilation -->
+      <ResolvedFileToPublish Include="@(IntermediateAssembly->'$(NativeOutputPath)%(Filename)$(NativeBinaryExt)')">
+        <RelativePath>%(Filename)$(NativeBinaryExt)</RelativePath>
+        <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+      </ResolvedFileToPublish>
     </ItemGroup>
 
-    <ItemGroup>
-      <_NativeIntermediateAssembly Include="@(IntermediateAssembly->'$(NativeOutputPath)%(Filename)$(NativeBinaryExt)')" />
-      <IntermediateAssembly Remove="@(IntermediateAssembly)" />
-      <IntermediateAssembly Include="@(_NativeIntermediateAssembly)" />
+    <PropertyGroup Condition="'$(CopyOutputSymbolsToPublishDirectory)' == 'true'">
+      <_symbolExt Condition="'$(OS)' == 'Windows_NT'">$(NativeSymbolExt)</_symbolExt>
+      <_symbolExt Condition="'$(OS)' != 'Windows_NT'">$(NativeBinaryExt)$(NativeSymbolExt)</_symbolExt>
+      <_symbolSourcePath>$(NativeOutputPath)$(TargetName)$(_symbolExt)</_symbolSourcePath>
+      <_symbolTargetPath>$(PublishDir)\$(TargetName)$(_symbolExt)</_symbolTargetPath>
+      <!-- If the symbol is a dSYM bundle, it's a folder not a file. -->
+      <_symbolIsFile Condition="'$(NativeSymbolExt)' == '.dSYM'">false</_symbolIsFile>
+      <_symbolIsFile Condition="'$(_symbolIsFile)' == ''">true</_symbolIsFile>
+    </PropertyGroup>
+
+    <ItemGroup Condition="'$(CopyOutputSymbolsToPublishDirectory)' == 'true'">
+      <_symbolRecursivePath Include="$(NativeOutputPath)$(TargetName)$(_symbolExt)/**/*" />
```

Comment:
I think it may be possible to simplify this -- I'll have to refresh my MSbuild knowledge to remember if it's easy to just track everything in a subdir.